File: modules/masks.js
// ============================================================
// MODULE: masks
// Mutually exclusive land system masks.
// Priority order: Anthro → Riparian → Core → Grass → Shrub → Open
// Each mask explicitly excludes all higher-priority classes.
// ============================================================
 
// Compute all 6 land system masks from indices and thresholds.
// Returns {anthro, riparian, core, grass, shrub, open}
exports.compute = function(idx, T) {

  var mask_anthro = idx.ndbi.gt(T.ANTHRO_NDBI)
    .or(idx.ndvi.lt(T.ANTHRO_NDVI_MAX));

  var mask_riparian = idx.ndmi.gt(T.RIPARIAN_NDMI)
    .and(idx.ndmi_dry.gt(T.RIPARIAN_NDMI_DRY))
    .and(idx.ndvi_dry.gt(T.RIPARIAN_NDVI_DRY))
    .and(mask_anthro.not());

  var mask_core = idx.ndvi_dry.gt(T.CORE_NDVI_DRY)
    .and(idx.ndmi.gt(T.CORE_NDMI))
    .and(mask_anthro.not())
    .and(mask_riparian.not());

  var mask_grass = idx.ndvi_dry.lt(T.GRASS_NDVI_DRY_MAX)
    .and(idx.ndvi_amp.gt(T.GRASS_AMP_MIN)
      .or(idx.ndmi_dry.lt(T.GRASS_NDMI_DRY_MAX)))
    .and(mask_anthro.not())
    .and(mask_riparian.not())
    .and(mask_core.not());

  var mask_shrub = idx.ndvi_dry.gte(T.SHRUB_NDVI_DRY_MIN)
    .and(idx.ndvi_dry.lt(T.SHRUB_NDVI_DRY_MAX))
    .and(idx.ndmi_dry.gte(T.SHRUB_NDMI_DRY_MIN))
    .and(idx.ndmi_dry.lt(T.SHRUB_NDMI_DRY_MAX))
    .and(idx.ndvi_amp.gte(T.SHRUB_AMP_MIN))
    .and(idx.ndvi_amp.lt(T.SHRUB_AMP_MAX))
    .and(mask_anthro.not())
    .and(mask_riparian.not())
    .and(mask_core.not())
    .and(mask_grass.not());

  // Open woodland split into drier/moister sub-populations
  var mask_open_dry = idx.ndvi_dry.gte(T.OPEN_NDVI_DRY_MIN)
    .and(idx.ndvi_dry.lt(T.OPEN_NDVI_DRY_MID))
    .and(idx.ndmi.gte(T.OPEN_NDMI_MIN))
    .and(idx.ndmi.lte(T.OPEN_NDMI_SPLIT_HIGH))
    .and(mask_anthro.not()).and(mask_riparian.not())
    .and(mask_core.not()).and(mask_grass.not())
    .and(mask_shrub.not());

  var mask_open_moist = idx.ndvi_dry.gte(T.OPEN_NDVI_DRY_MID)
    .and(idx.ndvi_dry.lte(T.OPEN_NDVI_DRY_MAX))
    .and(idx.ndmi.gt(T.OPEN_NDMI_SPLIT_LOW))
    .and(idx.ndmi.lte(T.OPEN_NDMI_MAX))
    .and(mask_anthro.not()).and(mask_riparian.not())
    .and(mask_core.not()).and(mask_grass.not())
    .and(mask_shrub.not());

  return {
    anthro:   mask_anthro,
    riparian: mask_riparian,
    core:     mask_core,
    grass:    mask_grass,
    shrub:    mask_shrub,
    open:     mask_open_dry.or(mask_open_moist)
  };
};


// ── Visualisation helpers ────────────────────────────────────
exports.addLayers = function(masks) {
  Map.addLayer(masks.core.selfMask(),
    {palette:['1a6b1a']}, 'MASK: Core Woodland', false);
  Map.addLayer(masks.open.selfMask(),
    {palette:['74c476']}, 'MASK: Open Woodland', false);
  Map.addLayer(masks.shrub.selfMask(),
    {palette:['addd8e']}, 'MASK: Shrub-Transition', false);
  Map.addLayer(masks.grass.selfMask(),
    {palette:['ffff99']}, 'MASK: Grassland', false);
  Map.addLayer(masks.riparian.selfMask(),
    {palette:['4292c6']}, 'MASK: Riparian', false);
  Map.addLayer(masks.anthro.selfMask(),
    {palette:['d73027']}, 'MASK: Anthropogenic', false);
};

exports.printCoverage = function(masks, geometry) {
  var check = function(mask, name) {
    print(name + ' coverage:', mask.unmask(0).reduceRegion({
      reducer:  ee.Reducer.mean(),
      geometry: geometry,
      scale:    100,
      maxPixels:1e9,
      tileScale:8
    }));
  };
  print('=== MASK COVERAGE FRACTIONS (target 0.05-0.30) ===');
  check(masks.core,     'Core Woodland');
  check(masks.open,     'Open Woodland');
  check(masks.shrub,    'Shrub-Transition');
  check(masks.grass,    'Grassland');
  check(masks.riparian, 'Riparian');
  check(masks.anthro,   'Anthropogenic');
};

File: savana/masks.py
"""Mutually exclusive land system masks.

Priority order: Anthropogenic -> Riparian -> Core -> Grassland ->
Shrub-Transition -> Open Woodland. Each mask explicitly excludes all
higher-priority classes. Direct port of ``masks.js``.
"""

from __future__ import annotations


def compute(idx: dict, T: dict) -> dict:
    """Compute all 6 land system masks from indices and thresholds.

    Returns ``{"anthro", "riparian", "core", "grass", "shrub", "open"}``.
    """
    mask_anthro = idx["ndbi"].gt(T["ANTHRO_NDBI"]).Or(
        idx["ndvi"].lt(T["ANTHRO_NDVI_MAX"])
    )

    mask_riparian = (
        idx["ndmi"]
        .gt(T["RIPARIAN_NDMI"])
        .And(idx["ndmi_dry"].gt(T["RIPARIAN_NDMI_DRY"]))
        .And(idx["ndvi_dry"].gt(T["RIPARIAN_NDVI_DRY"]))
        .And(mask_anthro.Not())
    )

    mask_core = (
        idx["ndvi_dry"]
        .gt(T["CORE_NDVI_DRY"])
        .And(idx["ndmi"].gt(T["CORE_NDMI"]))
        .And(mask_anthro.Not())
        .And(mask_riparian.Not())
    )

    mask_grass = (
        idx["ndvi_dry"]
        .lt(T["GRASS_NDVI_DRY_MAX"])
        .And(
            idx["ndvi_amp"]
            .gt(T["GRASS_AMP_MIN"])
            .Or(idx["ndmi_dry"].lt(T["GRASS_NDMI_DRY_MAX"]))
        )
        .And(mask_anthro.Not())
        .And(mask_riparian.Not())
        .And(mask_core.Not())
    )

    mask_shrub = (
        idx["ndvi_dry"]
        .gte(T["SHRUB_NDVI_DRY_MIN"])
        .And(idx["ndvi_dry"].lt(T["SHRUB_NDVI_DRY_MAX"]))
        .And(idx["ndmi_dry"].gte(T["SHRUB_NDMI_DRY_MIN"]))
        .And(idx["ndmi_dry"].lt(T["SHRUB_NDMI_DRY_MAX"]))
        .And(idx["ndvi_amp"].gte(T["SHRUB_AMP_MIN"]))
        .And(idx["ndvi_amp"].lt(T["SHRUB_AMP_MAX"]))
        .And(mask_anthro.Not())
        .And(mask_riparian.Not())
        .And(mask_core.Not())
        .And(mask_grass.Not())
    )

    mask_open_dry = (
        idx["ndvi_dry"]
        .gte(T["OPEN_NDVI_DRY_MIN"])
        .And(idx["ndvi_dry"].lt(T["OPEN_NDVI_DRY_MID"]))
        .And(idx["ndmi"].gte(T["OPEN_NDMI_MIN"]))
        .And(idx["ndmi"].lte(T["OPEN_NDMI_SPLIT_HIGH"]))
        .And(mask_anthro.Not())
        .And(mask_riparian.Not())
        .And(mask_core.Not())
        .And(mask_grass.Not())
        .And(mask_shrub.Not())
    )

    mask_open_moist = (
        idx["ndvi_dry"]
        .gte(T["OPEN_NDVI_DRY_MID"])
        .And(idx["ndvi_dry"].lte(T["OPEN_NDVI_DRY_MAX"]))
        .And(idx["ndmi"].gt(T["OPEN_NDMI_SPLIT_LOW"]))
        .And(idx["ndmi"].lte(T["OPEN_NDMI_MAX"]))
        .And(mask_anthro.Not())
        .And(mask_riparian.Not())
        .And(mask_core.Not())
        .And(mask_grass.Not())
        .And(mask_shrub.Not())
    )

    return {
        "anthro": mask_anthro,
        "riparian": mask_riparian,
        "core": mask_core,
        "grass": mask_grass,
        "shrub": mask_shrub,
        "open": mask_open_dry.Or(mask_open_moist),
    }


def coverage(masks: dict, region) -> dict:
    """Fraction of AOI covered by each mask (target range ~0.05-0.30 each)."""
    import ee

    out = {}
    for name, mask in masks.items():
        stats = mask.unmask(0).reduceRegion(
            reducer=ee.Reducer.mean(),
            geometry=region,
            scale=100,
            maxPixels=1e9,
            tileScale=8,
        )
        out[name] = stats
    return out


def add_layers(masks: dict, m=None, class_info: dict | None = None):
    """Add each mask as a layer to a geemap/folium Map (or ``ee.Map`` in Code Editor context).

    ``m`` should be a ``geemap.Map`` instance in a notebook. If omitted,
    a new one is created and returned.
    """
    from . import config

    if m is None:
        import geemap

        m = geemap.Map()

    info = class_info or config.DEFAULT_CLASS_INFO
    color_by_name = {
        "core": info[1]["color"],
        "open": info[2]["color"],
        "shrub": info[3]["color"],
        "grass": info[4]["color"],
        "riparian": info[5]["color"],
        "anthro": info[6]["color"],
    }
    labels = {
        "core": "MASK: Core Woodland",
        "open": "MASK: Open Woodland",
        "shrub": "MASK: Shrub-Transition",
        "grass": "MASK: Grassland",
        "riparian": "MASK: Riparian",
        "anthro": "MASK: Anthropogenic",
    }
    for key in ["core", "open", "shrub", "grass", "riparian", "anthro"]:
        m.addLayer(
            masks[key].selfMask(),
            {"palette": [color_by_name[key]]},
            labels[key],
            False,
        )
    return m
